rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -208,7 +207,9 @@ class Publisher : public PublisherBase
       if (RCL_RET_OK != status) {
         rclcpp::exceptions::throw_from_rcl_error(status, "failed to publish intra process message");
       }
+      this->do_inter_process_publish(msg_ptr);
```

Comment:
The function call `store_intra_process_message_()` above is taking ownership of the `msg_ptr`. So we can't rely on `msg_ptr` being still valid here. That is reason why the "inter" call currently happens *before* the "intra" call.

It would certainly be good to address the latency issue you describe but it will probably be more complicated than this.